backend/src/utils/actions.ts

```diff
@@ -64,103 +66,56 @@ export type Action =
-    access_key: {
-      nonce: number;
-      permission:
-        | {
-            permission_kind: "FUNCTION_CALL";
-            permission_details: {
-              allowance: string;
-              receiver_id: string;
-              method_names: string[];
-            };
-          }
-        | {
-            permission_kind: "FULL_ACCESS";
-          };
-    };
-  };
-};
+type DatabaseArgs =
+  Indexer.SelectorModelTypeMap["action_receipt_actions"]["args"];
 
-type DatabaseCreateAccount = {
-  kind: "CREATE_ACCOUNT";
-  args: {};
-};
+type DatabaseAddKeyArgs = Extract<DatabaseArgs, { access_key: unknown }>;
+type DatabaseCreateAccountArgs = Extract<DatabaseArgs, Record<string, never>>;
+type DatabaseDeleteAccountArgs = Extract<
+  DatabaseArgs,
+  { beneficiary_id: unknown }
+>;
+type DatabaseDeleteKeyArgs = Extract<
+  Exclude<DatabaseArgs, DatabaseAddKeyArgs | DatabaseStakeArgs>,
+  { public_key: unknown }
+>;
+type DatabaseDeployContractArgs = Extract<
+  DatabaseArgs,
+  { code_sha256: unknown }
+>;
+type DatabaseFunctionCallArgs = Extract<DatabaseArgs, { method_name: unknown }>;
+type DatabaseStakeArgs = Extract<DatabaseArgs, { stake: unknown }>;
+type DatabaseTransferArgs = Extract<
+  Exclude<DatabaseArgs, DatabaseFunctionCallArgs>,
+  { deposit: unknown }
+>;
 
-type DatabaseDeleteAccount = {
-  kind: "DELETE_ACCOUNT";
-  args: {
-    beneficiary_id: string;
-  };
+type DatabaseActionMapping = {
+  ADD_KEY: DatabaseAddKeyArgs;
+  CREATE_ACCOUNT: DatabaseCreateAccountArgs;
+  DELETE_ACCOUNT: DatabaseDeleteAccountArgs;
+  DELETE_KEY: DatabaseDeleteKeyArgs;
+  DEPLOY_CONTRACT: DatabaseDeployContractArgs;
+  FUNCTION_CALL: DatabaseFunctionCallArgs;
+  STAKE: DatabaseStakeArgs;
+  TRANSFER: DatabaseTransferArgs;
 };
 
-type DatabaseDeleteKey = {
-  kind: "DELETE_KEY";
-  args: {
-    public_key: string;
-  };
-};
+type ArgsTuple<Mapping extends DatabaseActionMapping = DatabaseActionMapping> =
+  {
+    [Kind in keyof Mapping]: [Kind, Mapping[Kind]];
+  }[keyof Mapping];
 
-type DatabaseDeployContract = {
-  kind: "DEPLOY_CONTRACT";
-  args: {
-    code_sha256: string;
-  };
-};
-
-type DatabaseFunctionCall = {
-  kind: "FUNCTION_CALL";
-  args: {
-    gas: number;
-    deposit: string;
-    method_name: string;
-    args_json?: Record<string, unknown>;
-    args_base64: string;
-  };
-};
-
-type DatabaseStake = {
-  kind: "STAKE";
-  args: {
-    public_key: string;
-    stake: string;
-  };
-};
-
-type DatabaseTransfer = {
-  kind: "TRANSFER";
-  args: {
-    deposit: string;
-  };
-};
-
-export type DatabaseAction = {
-  hash: string;
-} & (
-  | DatabaseAddKey
-  | DatabaseCreateAccount
-  | DatabaseDeleteAccount
-  | DatabaseDeleteKey
-  | DatabaseDeployContract
-  | DatabaseFunctionCall
-  | DatabaseStake
-  | DatabaseTransfer
-);
-
-export const mapDatabaseActionToAction = (action: DatabaseAction): Action => {
-  switch (action.kind) {
+const mapDatabaseActionToAction = (...[kind, args]: ArgsTuple): Action => {
```

Comment:
This is kinda messy, but it is a known request feature in [typescript](https://github.com/microsoft/TypeScript/issues/33014).